README.md
# Transcoder plugin for Craft CMS

Transcode videos to various formats, and provide thumbnails of the video

Related: [Transcoder for Craft 3.x](https://github.com/nystudio107/craft3-transcoder)

## Installation

To install Transcoder, follow these steps:

1. Download & unzip the file and place the `transcoder` directory into your `craft/plugins` directory
2.  -OR- do a `git clone https://github.com/nystudio107/transcoder.git` directly into your `craft/plugins` folder.  You can then update it with `git pull`
3.  -OR- install with Composer via `composer require nystudio107/transcoder`
4. Install plugin in the Craft Control Panel under Settings > Plugins
5. The plugin folder should be named `transcoder` for Craft to see it.  GitHub recently started appending `-master` (the branch name) to the name of the folder for zip file downloads.

Transcoder works on Craft 2.4.x, Craft 2.5.x and Craft 2.6.x.  `ffmpeg` **must** be installed for it to function.

## Transcoder Overview

The Transcoder video allows you to take any locally stored video, and transcode it into any bitrate/framerate, and save it out as a web-ready `.mp4` file.

It also allows you to get a thumbnail of the video in any size and at any timecode.

Finally, it lets you download an arbitrary file (such as the transcoded video) via a special download link.

If the source file has changed since the last time the video was encoded, it will re-encode the video and replace it.

## Configuring Transcoder

The only configuration for Transcoder is in the `config.php` file, which is a multi-environment friendly way to store the default settings.  Don't edit this file, instead copy it to `craft/config` as `transcoder.php` and make your changes there.  Here's the default `config.php` file:

    <?php
    /**
     * Transcoder Configuration
     *
     * Completely optional configuration settings for Transcoder if you want to customize some
     * of its more esoteric behavior, or just want specific control over things.
     *
     * Don't edit this file, instead copy it to 'craft/config' as 'transcoder.php' and make
     * your changes there.
     */

    return array(

    /**
     * The path to the ffmpeg binary
     */
        "ffmpegPath" => "/usr/bin/ffmpeg",

    /**
     * The path where the transcoded videos are stored
     */

        "transcoderPath" => $_SERVER['DOCUMENT_ROOT'] . "/transcoder/",

    /**
     * The URL where the transcoded videos are stored
     */

        "transcoderUrl" => "/transcoder/",
    );

## Using Transcoder

### Generating a Transcoded Video

To generate a transcoded video, do the following:

    {% set transVideoUrl = craft.transcoder.getVideoUrl('/home/vagrant/sites/nystudio107/public/trimurti.mp4', {
        "frameRate": 20,
        "bitRate": "500k"
    }) %}

You can also pass in an `AssetFileModel`:

    {% set myAsset = entry.someAsset.first() %}
    {% set transVideoUrl = craft.transcoder.getVideoUrl(myAsset, {
        "frameRate": 20,
        "bitRate": "500k"
    }) %}

It will return to you a URL to the transcoded video if it already exists, or if it doesn't exist, it will return `""` and kick off the transcoding process (which can be quite lengthy for long videos).

In the array you pass in, the default values are used if the key/value pair does not exist:

    {
        "bitRate" => "800k",
        "frameRate" => 15,
    }

If you want to have the Transcoder not change a parameter, pass in an empty value in the key/value pair, e.g.:

    {% set transVideoUrl = craft.transcoder.getVideoUrl('/home/vagrant/sites/nystudio107/public/trimurti.mp4', {
        "frameRate": "",
        "bitRate": ""
    }) %}

The above example would cause it to not change the frameRate or bitRate of the source movie (not recommended for client-proofing purposes).

### Generating a Video Thumbnail

To generate a thumbnail from a video, do the following:

    {% set transVideoThumbUrl = craft.transcoder.getVideoThumbnailUrl('/home/vagrant/sites/nystudio107/public/trimurti.mp4', {
        "width": 300,
        "height": 200,
        "timeInSecs": 20,
    }) %}

It will return to you a URL to the thumbnail of the video, in the size you specify, from the timecode `timeInSecs` in the video.  It creates this thumbnail immediately if it doesn't already exist.

In the array you pass in, the default values are used if the key/value pair does not exist:

    {
        "width" => 200,
        "height" => 100,
        "timeInSecs" => 10,
    }

If you want to have the Transcoder not change a parameter, pass in an empty value in the key/value pair, e.g.:

    {% set transVideoThumbUrl = craft.transcoder.getVideoThumbnailUrl('/home/vagrant/sites/nystudio107/public/trimurti.mp4', {
        "width": "",
        "height": "",
        "timeInSecs": 20,
    }) %}

The above example would cause it to generate a thumbnail at whatever size the video is (not recommended for client-proofing purposes).

### Generating a Download URL

To generate a download URL for a file, do the following:

    {% set downloadUrl = craft.transcoder.getDownloadUrl('/some/url') %}

When the user clicks on the URL, it will download the file to their local computer.  If the file doesn't exist, `""` is returned.

The file must reside in the webroot (thus a URL or URI must be passed in as a parameter, not a path), for security reasons.

## Transcoder Roadmap

Some things to do, and ideas for potential features:

* The videos could potentially be saved in different formats (though `.mp4` really is "the" standard for video)
* The videos could potentially be resized, either to an aspect ratio or an absolute size or what have you
* Accessors could be written to get information about a video (height, width, duration, and so on)

Brought to you by [nystudio107](https://nystudio107.com)

CHANGELOG.md
# Transcoder Changelog

## 1.1.1 - 2017.03.05
### Changed
* Fixed an issue with the downloadFile controller action
* Broke out the changelog to `CHANGELOG.md`
* Updated `README.md`

## 1.1.0 - 2016.09.12
### Added
* Initial release

Brought to you by [nystudio107](https://nystudio107.com)

TranscoderPlugin.php
<?php
/**
 * Transcoder plugin for Craft CMS
 *
 * Transcode videos to various formats, and provide thumbnails of the video
 *
 * @author    nystudio107
 * @copyright Copyright (c) 2016 nystudio107
 * @link      http://nystudio107.com
 * @package   Transcoder
 * @since     1.0.0
 */

namespace Craft;

class TranscoderPlugin extends BasePlugin
{
    /**
     * @return mixed
     */
    public function init()
    {
    }

    /**
     * @return mixed
     */
    public function getName()
    {
         return Craft::t('Transcoder');
    }

    /**
     * @return mixed
     */
    public function getDescription()
    {
        return Craft::t('Transcode videos to various formats, and provide thumbnails of the video');
    }

    /**
     * @return string
     */
    public function getVersion()
    {
        return '1.1.0';
    }

    /**
     * @return string
     */
    public function getSchemaVersion()
    {
        return '1.0.0';
    }

    /**
     * @return string
     */
    public function getDeveloper()
    {
        return 'nystudio107';
    }

    /**
     * @return string
     */
    public function getDeveloperUrl()
    {
        return 'https://nystudio107.com';
    }

    /**
     * @return string
     */
    public function getReleaseFeedUrl()
    {
        return 'https://raw.githubusercontent.com/nystudio107/transcoder/master/releases.json';
    }

    /**
     * @return string
     */
    public function getDocumentationUrl()
    {
        return 'https://github.com/nystudio107/transcoder/blob/master/README.md';
    }

    /**
     * @return bool
     */
    public function hasCpSection()
    {
        return false;
    }

    /**
     */
    public function onBeforeInstall()
    {
    }

    /**
     */
    public function onAfterInstall()
    {
    }

    /**
     */
    public function onBeforeUninstall()
    {
    }

    /**
     */
    public function onAfterUninstall()
    {
    }

}

variables/TranscoderVariable.php
<?php
/**
 * Transcoder plugin for Craft CMS
 *
 * Transcode videos to various formats, and provide thumbnails of the video
 *
 * @author    nystudio107
 * @copyright Copyright (c) 2016 nystudio107
 * @link      http://nystudio107.com
 * @package   Transcoder
 * @since     1.0.0
 */

namespace Craft;

class TranscoderVariable
{
    /**
     * Returns a URL to the transcoded video or "" if it doesn't exist (at which time it will create it)
     * By default, the video is never resized, and the format is always .mp4
     *
     * @param  string   $filepath       path to the original video
     * @param  array    $videoOptions   array of options for the video
     * @return string                   URL of the transcoded video or ""
     */
    public function getVideoUrl($filepath, $videoOptions)
    {
        $result = craft()->transcoder->getVideoUrl($filepath, $videoOptions);
        return $result;
    } /* -- getVideoUrl */

    /**
     * Returns a URL to a video thumbnail
     *
     * @param  string   $filepath           path to the original video
     * @param  array    $thumbnailOptions   array of options for the thumbnail
     * @return string                       URL of the video thumbnail
     */
    public function getVideoThumbnailUrl($filepath, $thumbnailOptions)
    {
        $result = craft()->transcoder->getVideoThumbnailUrl($filepath, $thumbnailOptions);
        return $result;
    } /* -- getVideoThumbnailUrl */

    /**
     * Get a download URL
     *
     * @param  string $url to the asset to download
     * @return string the download URL
     */
    function getDownloadUrl($url)
    {
        $result = "";
        $filepath = parse_url($url, PHP_URL_PATH);
        $filepath = $_SERVER['DOCUMENT_ROOT'] . $filepath;
        if (file_exists($filepath))
        {
            $urlParams = array(
                'url' => urlencode($url),
                );
            $result = UrlHelper::getActionUrl('transcoder/downloadFile', $urlParams);
        }
        return $result;
    } /* -- getDownloadUrl */

}

controllers/TranscoderController.php
<?php
/**
 * Transcoder plugin for Craft CMS
 *
 * Transcode videos to various formats, and provide thumbnails of the video
 *
 * @author    nystudio107
 * @copyright Copyright (c) 2016 nystudio107
 * @link      http://nystudio107.com
 * @package   Transcoder
 * @since     1.0.0
 */

namespace Craft;

class TranscoderController extends BaseController
{

    /**
     * @var    bool|array Allows anonymous access to this controller's actions.
     * @access protected
     */
    protected $allowAnonymous = array('actionDownloadFile',
        );

    /**
     * Force the download of a given $url.  We do it this way to prevent people from downloading
     * things that are outside of the server root.
     */
    public function actionDownloadFile()
    {
        $url = urldecode(craft()->request->getParam('url'));
        $filepath = parse_url($url, PHP_URL_PATH);
        $filepath = $_SERVER['DOCUMENT_ROOT'] . $filepath;
        $content = IOHelper::getFileContents($filepath);
        craft()->request->sendFile($filepath, $content, array('forceDownload' => true), true);
    } /* -- actionDownloadFile */

}

services/TranscoderService.php
<?php
/**
 * Transcoder plugin for Craft CMS
 *
 * Transcode videos to various formats, and provide thumbnails of the video
 *
 * @author    nystudio107
 * @copyright Copyright (c) 2016 nystudio107
 * @link      http://nystudio107.com
 * @package   Transcoder
 * @since     1.0.0
 */

namespace Craft;

class TranscoderService extends BaseApplicationComponent
{

    /**
     * Returns a URL to the transcoded video or "" if it doesn't exist (at which time it will create it)
     * By default, the video is never resized, and the format is always .mp4
     *
     * @param  string   $filepath       path to the original video -OR- an AssetFileModel
     * @param  array    $videoOptions   array of options for the video
     * @return string                   URL of the transcoded video or ""
     */
    public function getVideoUrl($filepath, $videoOptions)
    {

        $result = "";

/* -- If we're passed an AssetFileModel, extract the path from it */

        if (is_object($filepath))
        {
            $asset = $filepath;
            $source = $asset->getSource();

            if ($source->type != 'Local')
            {
                throw new Exception(Craft::t('Paths not available for non-local asset sources'));
            }

            $sourcePath = $source->settings['path'];
            $folderPath = $asset->getFolder()->path;

            $filepath = $sourcePath . $folderPath . $asset->filename;
        }

        if (file_exists($filepath))
        {
            $path_parts = pathinfo($filepath);
            $destVideoFile = $path_parts['filename'];
            $destVideoPath = craft()->config->get("transcoderPath", "transcoder");

/* -- Default options for transcoded videos */

            $defaultOptions = array(
                "bitRate" => "800k",
                "frameRate" => 15,
                );

/* -- Coalesce the passed in $videoOptions with the $defaultOptions */

            $videoOptions = array_merge($defaultOptions, $videoOptions);

/* -- Build the basic command for ffmpeg */

            $ffmpegCmd = craft()->config->get("ffmpegPath", "transcoder")
                . ' -i ' . escapeshellarg($filepath)
                . ' -vcodec libx264'
                . ' -vprofile high'
                . ' -preset slow'
                . ' -crf 22'
                . ' -c:a copy'
                . ' -bufsize 1000k'
                . ' -threads 0';

/* -- Set the framerate if desired */

            if ($videoOptions['frameRate'])
            {
                $ffmpegCmd .= ' -r ' . $videoOptions['frameRate'];
                $destVideoFile .= '_' . $videoOptions['frameRate'] . 'fps';
            }

/* -- Set the bitrate if desired */

            if ($videoOptions['bitRate'])
            {
                $ffmpegCmd .= ' -b:v ' . $videoOptions['bitRate'] . ' -maxrate ' . $videoOptions['bitRate'];
                $destVideoFile .= '_' . $videoOptions['bitRate'] . 'bps';
            }

/* -- Create the directory if it isn't there already */

            if (!file_exists($destVideoPath))
                mkdir($destVideoPath);

/* -- Assemble the destination path and final ffmpeg command */

            $destVideoFile .= '.mp4';
            $destVideoPath = $destVideoPath . $destVideoFile;
            $ffmpegCmd .= ' -f mp4 -y ' . escapeshellarg($destVideoPath) . ' >/dev/null 2>/dev/null & echo $!';

/* -- Make sure there isn't a lockfile for this video already */

            $oldpid = false;
            $lockfile = sys_get_temp_dir() . '/' . $destVideoFile . "lock";
            $oldpid = @file_get_contents($lockfile);
            if ($oldpid !== false)
            {
                exec("ps $oldpid", $ProcessState);
                if (count($ProcessState) >= 2)
                    return $result;
                unlink($lockfile);
            }

/* -- If the video file already exists and hasn't been modified, return it.  Otherwise, start it transcoding */

            if (file_exists($destVideoPath) && (filemtime($destVideoPath) >= filemtime($filepath)))
                $result = craft()->config->get("transcoderUrl", "transcoder") . $destVideoFile;
            else
            {

/* -- Kick off the transcoding */

                $pid = shell_exec($ffmpegCmd);
                TranscoderPlugin::log($ffmpegCmd, LogLevel::Info, false);

/* -- Create a lockfile in tmp */

                file_put_contents($lockfile, $pid);
            }
        }
        return $result;
    } /* -- getVideoUrl */

    /**
     * Returns a URL to a video thumbnail
     *
     * @param  string   $filepath           path to the original video or an AssetFileModel
     * @param  array    $thumbnailOptions   array of options for the thumbnail
     * @return string                       URL of the video thumbnail
     */
    public function getVideoThumbnailUrl($filepath, $thumbnailOptions)
    {

        $result = "";

/* -- If we're passed an AssetFileModel, extract the path from it */

        if (is_object($filepath))
        {
            $asset = $filepath;
            $source = $asset->getSource();

            if ($source->type != 'Local')
            {
                throw new Exception(Craft::t('Paths not available for non-local asset sources'));
            }

            $sourcePath = $source->settings['path'];
            $folderPath = $asset->getFolder()->path;

            $filepath = $sourcePath . $folderPath . $asset->filename;
        }

        if (file_exists($filepath))
        {
            $path_parts = pathinfo($filepath);
            $destThumbnailFile = $path_parts['filename'];
            $destThumbnailPath = craft()->config->get("transcoderPath", "transcoder");

/* -- Default options for video thumbnails */

            $defaultOptions = array(
                "width" => 200,
                "height" => 100,
                "timeInSecs" => 10,
                );

/* -- Coalesce the passed in $thumbnailOptions with the $defaultOptions */

            $thumbnailOptions = array_merge($defaultOptions, $thumbnailOptions);

/* -- Build the basic command for ffmpeg */

            $ffmpegCmd = craft()->config->get("ffmpegPath", "transcoder")
                . ' -i ' . escapeshellarg($filepath)
                . ' -vcodec mjpeg'
                . ' -vframes 1';

/* -- Set the width & height if desired */

            if ($thumbnailOptions['width'] && $thumbnailOptions['height'])
            {
                $ffmpegCmd .= ' -vf "scale='
                    . $thumbnailOptions['width'] . ':' . $thumbnailOptions['height']
                    . ', unsharp=5:5:1.0:5:5:0.0"';
                $destThumbnailFile .= '_' . $thumbnailOptions['width'] . 'x' . $thumbnailOptions['height'];
            }

/* -- Set the timecode to get the thumbnail from if desired */

            if ($thumbnailOptions['timeInSecs'])
            {
                $timecode = gmdate("H:i:s", $thumbnailOptions['timeInSecs']);
                $ffmpegCmd .= ' -ss ' . $timecode . '.00';
                $destThumbnailFile .= '_' . $thumbnailOptions['timeInSecs'] . 's';
            }

/* -- Create the directory if it isn't there already */

            if (!file_exists($destThumbnailPath))
                mkdir($destThumbnailPath);

/* -- Assemble the destination path and final ffmpeg command */

            $destThumbnailFile .= '.jpg';
            $destThumbnailPath = $destThumbnailPath . $destThumbnailFile;
            $ffmpegCmd .= ' -f image2 -y ' . escapeshellarg($destThumbnailPath) . ' >/dev/null 2>/dev/null';

/* -- If the thumbnail file already exists, return it.  Otherwise, generate it and return it */

            if (file_exists($destThumbnailPath))
                $result = craft()->config->get("transcoderUrl", "transcoder") . $destThumbnailFile;
            else
            {
                $pid = shell_exec($ffmpegCmd);
                TranscoderPlugin::log($ffmpegCmd, LogLevel::Info, false);
                $result = craft()->config->get("transcoderUrl", "transcoder") . $destThumbnailFile;
            }
        }
        return $result;
    } /* -- getVideoThumbnailUrl */

}